cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -905,6 +905,56 @@ private void addWorkflowNodeState(ProgramRunId programRunId, Map<String, String>
     }
   }
 
+  /**
+   * Record that the program run is pending run.
+   *
+   * @param programRunId program run
+   *
+   * @return {@link RunRecordDetail} with status Pending.
+   */
+  @Nullable
+  public RunRecordDetail recordProgramPending(ProgramRunId programRunId, Map<String, String> runtimeArgs,
+      Map<String, String> systemArgs, @Nullable ArtifactId artifactId)
+      throws IOException {
+    long startTs = RunIds.getTime(programRunId.getRun(), TimeUnit.SECONDS);
+    if (startTs == -1L) {
+      LOG.error(
```

Comment:
I would fail it. Since we are in the handler, we can fail and pass it to user. We could not do it in the subscriber

Reply:
Done.

Throwing `IllegalArgumentException` for this case.